### PR TITLE
feat(spooler): Add CachingEnvelopeStack

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR implements a new `CachingEnvelopeStack` which caches one element in memory before delegating to the inner `EnvelopeStack` implementation.

Closes: https://github.com/getsentry/team-ingest/issues/586

### CODE REVIEW

File: relay-server/src/services/buffer/stack_provider/sqlite.rs [hunk 1]
```diff
@@ -38,7 +39,7 @@ impl SqliteStackProvider {
 }
 
 impl StackProvider for SqliteStackProvider {
-    type Stack = SqliteEnvelopeStack;
+    type Stack = CachingEnvelopeStack<SqliteEnvelopeStack>;
```
Comment (iambriccardo) [inline relay-server/src/services/buffer/stack_provider/sqlite.rs:42]: For now, we want to directly use it with the `SqliteStackProvider`.

File: relay-server/src/services/buffer/envelope_stack/caching.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            self.cached = Some(envelope);
+            Ok(())
+        } else {
+            // If we already have a cached envelope, push the current cached one
+            // to the inner stack and keep the new one in cache
+            let cached = self.cached.take().unwrap();
+            self.inner.push(cached).await?;
+            self.cached = Some(envelope);
+            Ok(())
+        }
```
Comment (jjbayer) [inline relay-server/src/services/buffer/envelope_stack/caching.rs]:
```suggestion
        if let Some(cached) = self.cached.take() {
            self.inner.push(cached).await?;
        }
        self.cached = Some(envelope);
        Ok(())
```

File: relay-server/src/services/buffer/envelope_stack/caching.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        if let Some(envelope) = self.cached.take() {
+            Ok(Some(envelope))
+        } else {
+            self.inner.pop().await
+        }
+    }
+
+    async fn flush(mut self) {
+        if let Some(envelope) = self.cached {
+            let _ = self.inner.push(envelope).await;
```
Comment (jjbayer) [inline relay-server/src/services/buffer/envelope_stack/caching.rs]: Should we log an error here?

File: relay-server/src/services/buffer/envelope_stack/caching.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+        // Create test envelopes with different timestamps
+        let env1 = mock_envelope(Utc::now());
+        let env2 = mock_envelope(Utc::now());
+        let env3 = mock_envelope(Utc::now());
+
+        // Push envelopes
+        stack.push(env1).await.unwrap();
+        stack.push(env2).await.unwrap();
+        stack.push(env3).await.unwrap();
```
Comment (jjbayer) [inline relay-server/src/services/buffer/envelope_stack/caching.rs]: Can we test `stack.pop()` before pushing, and `peek()` before and in between push calls?

File: relay-server/src/services/buffer/envelope_stack/caching.rs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            Ok(())
+        }
+    }
+
+    async fn peek(&mut self) -> Result<Option<DateTime<Utc>>, Self::Error> {
+        if let Some(ref envelope) = self.cached {
+            Ok(Some(envelope.received_at()))
+        } else {
+            self.inner.peek().await
+        }
```
Comment (jjbayer) [inline relay-server/src/services/buffer/envelope_stack/caching.rs:49]: Tricky question whether peek should pop from the inner. Probably not.

**Review by jjbayer — APPROVED** [review]
Nice! Couple of nit picks.

File: relay-server/src/services/buffer/envelope_stack/caching.rs (hunk 5 above)
Comment (iambriccardo) [reply relay-server/src/services/buffer/envelope_stack/caching.rs:55]: Isn't the `peek` behavior supposed always to return a reference if there is at least one envelope? Your idea seems to break this invariant.

File: relay-server/src/services/buffer/envelope_stack/caching.rs (hunk 3 above)
Comment (iambriccardo) [reply relay-server/src/services/buffer/envelope_stack/caching.rs]: Yeah, maybe "Cached envelope" was not flushed to the inner stack.

**Pull request merged**